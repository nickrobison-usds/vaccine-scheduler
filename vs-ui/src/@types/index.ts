import Client from "fhir-kit-client";

export interface OperationParams<T extends fhir.DomainResource> {
    name: string;
    resourceType?: string;
    id?: string;
    method?: "get" | "post";
    input?: string | T;
}

export interface FhirClientResponse {
    status: number;
    data: fhir.OperationOutcome;
}

export interface FhirClientError {
    config: any; // Type this
    response: FhirClientResponse;
}

export interface OperationClient extends Client {
    operation<T extends fhir.DomainResource>(params: OperationParams<T>): Promise<fhir.OperationOutcome>;
}

// Guards

export function isClientError(err: any): err is FhirClientError {
    return (err as FhirClientError).response !== undefined;
}

export function isFHIRResource(json: any): json is fhir.DomainResource {
    return (json as fhir.DomainResource).resourceType !== undefined;
}

export function isBundle(bundle: any): bundle is fhir.Bundle {
    return (bundle as fhir.Bundle).type !== undefined;
}
