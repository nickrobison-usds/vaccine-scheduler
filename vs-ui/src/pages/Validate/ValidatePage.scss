.validate-header {
  text-align: center;
}
