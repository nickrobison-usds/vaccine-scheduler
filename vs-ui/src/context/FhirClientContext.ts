import { createContext, useContext } from "react";

import { OperationClient } from "../@types";

export type FhirContextType = {
  client?: OperationClient;
};

export const FhirClientContext = createContext<FhirContextType>({});
export const useFhir = () => useContext(FhirClientContext);
