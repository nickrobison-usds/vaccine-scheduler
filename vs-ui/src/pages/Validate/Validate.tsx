import {Alert, Button, ErrorMessage, Grid, Label, Textarea} from "@trussworks/react-uswds";
import {Field, Form, FormikBag, FormikProps, withFormik} from "formik";
import React, {useState} from "react";
import {useFhir} from "../../context/FhirClientContext";
import {isClientError, isFHIRResource} from "../../@types";

interface FormValues {
    text: string
}

const severityToAlertType = (severity: fhir.IssueSeverity): "error" | "warning" | "success" | "info" => {
    switch (severity) {
        case "error":
            return "error";
        default:
            return "error"
    }
}

const issueToAlert = (issue: fhir.OperationOutcomeIssue) => {
    return (
        <Alert type={severityToAlertType(issue.severity)} slim noIcon>
            {issue.diagnostics}
        </Alert>
    )
}

const ValidationForm = (props: FormikProps<FormValues>) => {
    const {
        touched,
        errors,
        handleSubmit,
        isSubmitting
    } = props;
    return (<Form onSubmit={handleSubmit} className="usa-form">
        <Label htmlFor="text">Paste text here:</Label>
        <Field id="text" as={Textarea} name="text" placeholder="Paste data here" aria-label="validation-text"/>
        {touched.text && errors.text && <ErrorMessage>{errors.text}</ErrorMessage>}
        <Button type="submit" disabled={isSubmitting}>Validate</Button>
    </Form>)
}

export const Validate: React.FC<{}> = () => {
    const {client} = useFhir();
    const [validationErrors, setValidationErrors] = useState<fhir.OperationOutcomeIssue[]>([]);

    const handleSubmission = async (values: FormValues, bag: FormikBag<any, any>) => {
        const {setFieldError} = bag;
        console.debug("I have values: ", values);

        // JSON stringify
        const maybeFHIR = JSON.parse(values.text);

        if (!isFHIRResource(maybeFHIR)) {
            // We need to handle this better
            setFieldError('text', 'Not a fhir resource');
            return;
        }
        try {
            const res = await client?.operation({
                name: 'validate',
                resourceType: 'Slot',
                input: maybeFHIR
            });
            console.debug("Response: ", res);
        } catch (exn) {
            if (isClientError(exn)) {
                console.error("Validation failed: ", exn.response.data);
                setValidationErrors(exn.response.data.issue);
            } else {
                console.error("Something else went wrong: ", exn);
            }
        }
    };

    const FormikForm = withFormik({
        displayName: 'ValidationForm',
        handleSubmit: handleSubmission,
    })(ValidationForm);

    return (
        <Grid>
            <Grid row={true}>
                <FormikForm/>
            </Grid>
            <Grid row={true}>
                <section className="usa-section">
                    {validationErrors.map(issueToAlert)}
                </section>
            </Grid>
        </Grid>
    )
}
