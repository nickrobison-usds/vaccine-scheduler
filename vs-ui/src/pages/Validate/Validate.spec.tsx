// noinspection HttpUrlsUsage

import { act, fireEvent, render, screen } from "@testing-library/react";
import Client from "fhir-kit-client";
import userEvent from "@testing-library/user-event";

import { FhirClientContext } from "../../context/FhirClientContext";
import { FhirClientError, OperationClient } from "../../@types";

import { Validate } from "./Validate";

jest.mock("fhir-kit-client");

const slotFailure: FhirClientError = {
  config: undefined,
  response: {
    status: 422,
    data: {
      resourceType: "OperationOutcome",
      issue: [
        {
          severity: "error",
          code: "processing",
          diagnostics:
            "Slot.schedule: minimum required = 1, but only found 0 (from http://fhir-registry.smarthealthit.org/StructureDefinition/vaccine-slot)",
          location: ["Slot", "Line 1, Col 35"],
        },
        {
          severity: "error",
          code: "processing",
          diagnostics:
            "Slot.status: minimum required = 1, but only found 0 (from http://fhir-registry.smarthealthit.org/StructureDefinition/vaccine-slot)",
          location: ["Slot", "Line 1, Col 35"],
        },
        {
          severity: "error",
          code: "processing",
          diagnostics:
            "Slot.start: minimum required = 1, but only found 0 (from http://fhir-registry.smarthealthit.org/StructureDefinition/vaccine-slot)",
          location: ["Slot", "Line 1, Col 35",
        },
        {
          severity: "error",
          code: "processing",
          diagnostics:
            "Slot.end: minimum required = 1, but only found 0 (from http://fhir-registry.smarthealthit.org/StructureDefinition/vaccine-slot)",
          location: ["Slot", "Line 1, Col 35",
        ,
      ],
    },
  },
};

describe("Validate", () => {
  it("renders text", async () => {
    const { container } = render(<Validate />);
    expect(screen.getByText(/text here/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("handles non-FHIR JSON", async () => {
    render(<Validate />);
    act(() => {
      fireEvent.input(screen.getByLabelText("validation-text"), {
        target: {
          value: '{"resz": "Slot", "id": 1}',
        },
      });
    });

    act(() => {
      fireEvent.submit(screen.getByRole("button", { name: /validate/i }));
    });
    expect(await screen.findByText("Not a fhir resource")).toBeInTheDocument();
  });

  it("requires valid json", async () => {
    render(<Validate />);
    expect(screen.getByRole("button", { name: /validate/i })).toBeDisabled();
    await userEvent.type(screen.getByLabelText("validation-text"), "hello");
    await userEvent.click(screen.getByRole("button", { name: /validate/i }));
    expect(screen.getByRole("button", { name: /validate/i })).toBeDisabled();
    expect(
      await screen.findByText("Data must be valid JSON")
    ).toBeInTheDocument();
  });

  it("displays validation errors", async () => {
    // @ts-ignore
    Client.mockImplementationOnce(() => {
      return {
        operation: () => Promise.reject(slotFailure),
      };
    });
    const client = new Client({ baseUrl: "http://nothing" }) as OperationClient;
    const { container } = render(
      <FhirClientContext.Provider value={{ client }}>
        <Validate />
      </FhirClientContext.Provider>
    );

    act(() => {
      fireEvent.input(screen.getByLabelText("validation-text"), {
        target: {
          value: '{"resourceType": "Slot", "id": 1}',
        },
      });
    });

    act(() => {
      fireEvent.submit(screen.getByRole("button", { name: /validate/i }));
    });
    expect(container).toMatchSnapshot();
  });
});
