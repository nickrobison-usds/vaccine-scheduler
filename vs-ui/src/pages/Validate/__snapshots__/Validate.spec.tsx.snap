// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate displays validation errors 1`] = `
<div>
  <div
    class="grid-container"
    data-testid="gridContainer"
  >
    <div
      class=""
      data-testid="grid"
    >
      <div
        class="grid-row"
        data-testid="grid"
      >
        <header
          class="usa-header validate-header"
          data-testid="header"
        >
          <h1
            id="validate-header"
          >
            Validate FHIR resources
          </h1>
        </header>
      </div>
      <div
        class="grid-row"
        data-testid="grid"
      >
        <form
          action="#"
          class="usa-form"
        >
          <label
            class="usa-label"
            data-testid="label"
            for="text"
          >
            Paste text here:
          </label>
          <textarea
            aria-label="validation-text"
            class="usa-textarea"
            data-testid="textarea"
            id="text"
            name="text"
            placeholder="Paste data here"
          >
            {"resourceType": "Slot", "id": 1}
          </textarea>
          <button
            class="usa-button"
            data-testid="button"
            disabled=""
            type="submit"
          >
            Validate
          </button>
        </form>
      </div>
      <div
        class="grid-row"
        data-testid="grid"
      >
        <section
          class="usa-section"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Validate renders text 1`] = `
<div>
  <div
    class="grid-container"
    data-testid="gridContainer"
  >
    <div
      class=""
      data-testid="grid"
    >
      <div
        class="grid-row"
        data-testid="grid"
      >
        <header
          class="usa-header validate-header"
          data-testid="header"
        >
          <h1
            id="validate-header"
          >
            Validate FHIR resources
          </h1>
        </header>
      </div>
      <div
        class="grid-row"
        data-testid="grid"
      >
        <form
          action="#"
          class="usa-form"
        >
          <label
            class="usa-label"
            data-testid="label"
            for="text"
          >
            Paste text here:
          </label>
          <textarea
            aria-label="validation-text"
            class="usa-textarea"
            data-testid="textarea"
            id="text"
            name="text"
            placeholder="Paste data here"
          />
          <button
            class="usa-button"
            data-testid="button"
            disabled=""
            type="submit"
          >
            Validate
          </button>
        </form>
      </div>
      <div
        class="grid-row"
        data-testid="grid"
      >
        <section
          class="usa-section"
        />
      </div>
    </div>
  </div>
</div>
`;
